### ISSUE TITLE
[ENHANCEMENT] Knuth & Plass line breaking algorithm

### ISSUE THREAD

**Issue by basil** [danielgarthur/neanes#90]
See [_Breaking Paragraphs into Lines_](http://www.eprg.org/G53DOC/pdfs/knuth-plass-breaking.pdf) by Donald Knuth & Michael Plass (1981). Serious typesetting programs that deal with justification, including TeX and Adobe InDesign, use a sophisticated algorithm for breaking paragraphs into lines. In the abovementioned paper, Knuth & Plass present a model of boxes, glue, and penalty specifications to efficiently compute the optimal set of line breaks. The Adobe Paragraph Composer works in much the same way. This avoids excessive hyphenation and rivers of white space in the justified output.

I have never seen anyone attempt to apply this concept to Western notation. About ten years ago I was using Finale, and I had to manually break the score into lines to get good output. I have never used Dorico and have no idea if it is any better.

Applying Knuth & Plass line breaking to Western notation music seems fairly difficult due to the geometric placement of notes on the Western staff, but it seems plausible that this algorithm could be applied to Byzantine notation music. The goal would be to avoid the bad line and page breaks that plague poorly typeset music. For example, most sticheraric and slow heirmologic pieces have a duple rhythm of 4-beat or 6-beat measures. Naïvely breaking the line between an odd-numbered beat and an even-numbered beat is undesirable, so this could be given a worse penalty specification in Knuth & Plass's algorithm. Similarly, a gorgon applies to the neume it is placed on as well as the preceding neume, so a line or page break between these two neumes is also undesirable.

**Comment by basil** [comment]
The very clever Romanian printer who typeset Anton Pann's books in the 19th century would occasionally use a slightly shorter ison or oligon in order to compress a line rather than expanding it (with the result of ugly rivers of white).

**Comment by basil** [comment]
Another thing this very clever Romanian printer does is this: when he is forced to justify a line with a lot of whitespace, he adds more space before and after a martyria than he does before and after the other notes of the line. This allows him to justify the line while still keeping the spacing between neumes more consistent with other lines.

**Comment by basil** [comment]
Both techniques in full display in this September 14 Megalynarion, from Anton Pann's book of Cherubic Hymns and Communion Hymns:

![September_14_Megalynarion](https://github.com/danielgarthur/neanes/assets/29850/4d99c3f4-ed69-40ee-992f-e61db7370c15)

**Comment by basil** [comment]
To translate this into Knuth & Plass' paradigm, the _glue_ (whitespace) around a martyria could have greater _stretchability_ than the glue around a regular neume.

Allowing for a shorter oligon or ison to be substituted during the breakpoint search could have highly beneficial effects with regard to computing the optimal breakpoints, but it would be tough to model in Knuth & Plass' paradigm and might also look inconsistent in a modern digitally-typeset score.

**Comment by basil** [comment]
I spent some time this week cleaning up my patches from this past December and am getting close to a point where I can submit this feature as a PR. I still have a few bugs left to fix and some areas to clean up, but I have tested the code on all my scores and I am quite pleased with the results. I believe they meet and exceed the quality of the 19th-century classical publications.

@danielgarthur I wanted to gather feedback on two areas as I get ready to submit this feature.

1. My code at https://github.com/basil/neanes/tree/knuth-plass replaces the first-fit algorithm with the optimum-fit algorithm from the paper by Knuth and Plass. There is no way to restore the old algorithm. My experience is that the new algorithm produces superior results in every case I have tested. I am OK with such a breaking change, since we don't currently make any commitment that old scores will look exactly the same in newer versions of the software. If you are OK with this breaking change, I would prefer to keep the code simple by not trying to support both algorithms. But if you find any examples where the new algorithm is worse than the old, then I would like to hear about them.
2. I have written a rather large essay about this change at https://github.com/basil/neanes/blob/knuth-plass/notes/knuth_plass.md, which I am sharing now in advance of the final PR because this is a lot of material to read and digest. I am interested in any early feedback on the essay and am happy to incorporate any high-level suggestions as I get closer to submitting a PR.

**Comment by danielgarthur** [comment]
I have read the essay and it sounds good. I have only just begun to look through the code, but I will continue to do so.

**Comment by basil** [comment]
I updated my branch tonight fixing most known bugs, but one bug remains. The Gourlay-style spacers are working well within a line, but not at the beginning and ends of lines. I need to split them apart into two halves, one per lyric, and model that in the box-glue algebra. I think I even wrote about it that way in the design document, but never actually implemented it back in December. That's the last portion I know of that needs to be finished.

**Comment by basil** [comment]
This has proven more difficult than expected, because modeling lyric spacing in the Knuth-Plass algebra requires the use of negative widths and/or negative stretchability, which the library I am using (https://github.com/robertknight/tex-linebreak) does not support. And Knuth does not really describe how to implement it in his paper, although it is implemented in TeX—though the TeX implementation is even more unreadable than the paper due to an even greater level of optimization. I guess I will start by trying to write a test case for negative stretchability in `tex-linebreak`, based on the centered text example in Knuth's paper. Then I will try to figure out how to enhance `tex-linebreak` to support this use case. My sense is that the restriction on negative widths is meant to support some of the optimizations in Knuth's paper (which, however, are not as deep as the optimizations in TeX), but also these optimizations might not be needed on modern hardware. So perhaps `tex-linebreak` can be simplified by rolling back some of these optimizations, at which point it might also be able to support negative widths, after which I can get on with modeling lyric spacing with it.

**Comment by basil** [comment]
> This has proven more difficult than expected, because modeling lyric spacing in the Knuth-Plass algebra requires the use of negative widths and/or negative stretchability, which the library I am using (https://github.com/robertknight/tex-linebreak) does not support.

I fixed this in the latest release of `tex-linebreak`. It turns out that the restriction on negative values were merely some performance optimizations added by Knuth to deal with the limitations of 1970s era hardware. These performance optimizations can be removed on modern hardware, which allows full support for negative values.

Supporting lyrics in Gourlay-style spacers continues to be difficult. For reference, the prior work on applying Knuth-Plass to music spacing is:

- John S Gourlay's original work, published in the 1987 papers _Spacing a Line of Music_ and _Optimal line breaking in music_.
- Lippold Haken and Dorothea Blostein's development of this concept in _[Justification of Printed Music](https://dl.acm.org/doi/pdf/10.1145/102868.102874)_ (1991) and _[A New Algorithm for Horizontal Spacing of Printed Music](https://quod.lib.umich.edu/cgi/p/pod/dod-idx/new-algorithm-for-horizontal-spacing-of-printed-music.pdf?c=icmc;idno=bbp2372.1995.032;format=pdf)_ (1995)
- Kai Renz's development of this concept in Guido in _[An improved algorithm for spacing a line of music](https://quod.lib.umich.edu/cgi/p/pod/dod-idx/improved-algorithm-for-spacing-a-line-of-music.pdf?c=icmc;idno=bbp2372.2002.097;format=pdf)_ (2002), which is a summary of his [PhD dissertation](https://github.com/grame-cncm/guidolib/blob/dev/doc/papers/kai_renz_diss.pdf) (well worth reading).
- A [confusing paper](http://jim.afim-asso.org/jim11/html/actes/28_40_paper.pdf) about applying this concept to "graphical notation" from 2011.

The trail in academia seems to leave off there, with this appearing to be more or less a solved problem for Western notation at this point. Gourlay style breaking is implemened in LilyPond. There are no references to Gourlay or Knuth in the MuseScore source code, and it's unclear what is used there. [This forum post](https://forums.steinberg.net/t/casting-off/670560/3) by Daniel Spreadbury indicates that Dorico uses a simplified version of Gourlay spacing, open to improvement but "it’s a non-trivial problem (to say the least)". 😅 

Several of the above papers make good observations about how to extend Gourlay's idea of spacers to work across multiple parts for polyphonic music, but none seem to deal with the problem of lyrics. It appears most of the above systems were focused on instrumental music.

it appears that the current maintainer of Guido (Dominique Fober) extended it in February of 2017 to add lyrics collision detection support to Kai Renz's implementation of the concept. This feature did come with tests, visible in https://guidoeditor.grame.fr/ in the "lyrics" section, but I can't seem to get the checkbox to do anything, so maybe it has regressed. Dominique did not write any academic papers about this most recent development, but the code is open source, so it might be possible to study the comments to learn from his approach. It also might be possible to contact Dominique and see if he could summarize, in an informal setting, his key insights in this space.

**Comment by basil** [comment]
I asked GPT 5 to summarize the Guido code by Dominique Fober to adapt Gourlay (Knuth-Plass) spacing to deal with lyric collisions. As far as I can tell, this is the only open source Gourlay implementation that attempts to deal with lyric collisions. Here is the summary.

---

After a first, normal Gourlay spacing pass, the engine scans the laid‑out score for lyric items whose bounding boxes overlap. For each collision, it computes how much extra horizontal space is needed and injects an automatic space tag (ARSpace) at the corresponding musical note in the abstract representation. Then it rebuilds the graphical representation so the spacing model (springs and rods) enforces these extra gaps.

How it works in the code
1) First layout using Gourlay’s spring–rod model:
- `GRMusic::createGR` builds voices and pages and delegates spacing to `GRStaffManager` / `GRSystem`.
- Spacing itself uses:
  - Springs (`GRSpring`) with constants derived from durations (see GRSpring::defconst, calcconst).
  - Rods (`GRRod`) to enforce minimum distances between consecutive springs (noteheads, bars, glues, etc.).
  - A space‑force function (`GRSpaceForceFunction2`) to find a force that stretches the spring set to the target width.
  - Optionally, a spacing matrix and neighborhood equalization (`GRSpacingMatrix`, `GRStaffManager::BuildSFF` with `kIsGiesekingSpacing` and `neighborhoodSpacing`) following Renz (2002).

2) Detect lyric collisions after that first layout:
- Entry point: `GRMusic::checkLyricsCollisions()`.
  - It calls `checkCollisions(true)`, which delegates to each page and then to each system (`GRSystem::checkCollisions`).
  - In the “lyrics” path, `GRSystem` gathers all lyric graphics per staff, then:
    - Computes their bounding boxes in device units, shifted to global positions (`GRSystem::checkCollisions(TCollisions&, std::vector<const GRNotationElement*>&)`).
    - For consecutive lyrics at the same musical note, it tests overlap; when a collision is detected, it computes a required extra space. The helper `checkCollision` returns approximately `(width1 + width2 + LSPACE/2)/2`, which creates a minimum margin (LSPACE/2) between lyric boxes.
    - It tells a TCollisions accumulator (`state.resolve(...`)) “at this note, we need at least this much extra space.”
- Back in `GRMusic`:
  - Collisions are sorted by musical note and consolidated (strip keeps, for a given note, only the largest requested gap).
  - `resolveCollision`s turns each consolidated request into an AR space:
    - It finds the AR voice and the AR object at that note, and inserts an `ARSpace` tag right after it (or after the barline if the object is a bar).
    - If the reference element is a position-tag (e.g., when lyrics are attached as position tags), it finds the next note and inserts the `ARSpace` just before it.
    - These `ARSpace` instances are flagged “auto” (getIsAuto() true) so they can be later removed.

3) Re‑layout to enforce the extra space using the spring–rod machinery:
- After insertion, `GRMusic::resolveCollisions` calls `createGR()` again.
- During `GRVoice::createNewRods`, GRSpace items are seen in the event stream:
  - They are converted into “space rods” (`GRRod` built with a fixed length equal to the requested `ARSpace` value, and `setIsSpaceRod(true)`).
  - `addRod(..., spaceactive=true)` freezes the spanned springs and adds the rod (`GRStaffManager::addRod`), i.e., it injects a hard local constraint into the initial stretch.
  - In `InitialSpringStretch`, space rods are applied with priority (`GRForceRodEntry::comp` sorts space rods first, and by earliest spring), so their fixed lengths are honored before the global stretch. This is exactly the Gourlay approach: add rods to impose local minimum distances, then solve the spring system globally.
- The second layout thus respects the added gaps and removes the collisions.

4) Toggling the feature off: remove auto space
- `GRMusic::removeAutoSpace(ARMusic*)`:
  - Iterates all AR voices and removes ARSpace whose getIsAuto() is true.
  - This is invoked automatically when settings are updated without `checkLyricsCollisions` (see `GuidoUpdateGR` in `GUIDOEngine.cpp`): if the last layout inserted auto spaces, they’re stripped before rebuilding the GR.
- After removal, a normal spring–rod layout (no extra space rods) is computed again.

Why this fits Gourlay / Haken–Blostein / Renz

- Gourlay (1987) models spacing as springs with rods imposing minimum distances. Here, lyric conflicts are resolved by adding rods (via ARSpace) that enforce extra room locally.
- Haken & Blostein (1991/1995) use a space‑force function. The implementation (GRSpaceForceFunction2) computes the force that maps to the required extent. Space rods are accounted for in the initial stretch, and frozen springs (from space rods) are correctly handled in the SFF.
- Renz (2002) adds a spacing matrix and neighborhood controls. Those are orthogonal and still applied (when enabled) before the lyric collision pass; the collision pass then corrects local exceptions by inserting constraint rods.

Notes and corner cases
- Collisions are resolved per musical note and merged to a single request (largest gap) to avoid over‑insertion of space when multiple staves collide at the same instant (`GRMusic::strip()`).
- Bars: if the collision reference AR object is a barline, the space is placed after it to avoid moving the bar itself.
- Minimum lyric margin: `checkCollision` enforces at least `LSPACE/2` clearance.
- The `ARSpace` auto flag is essential to distinguish user‑authored space from automatically inserted space when cleaning up (`removeAutoSpace`).

Key functions to look at
- `GRMusic::checkLyricsCollisions, checkCollisions, resolveCollisions, removeAutoSpace`
- `GRSystem::checkCollisions(state, lyrics) and helpers checkCollision`
- `GRVoice::createNewRods` (handling `GRSpace` into space rods)
- `GRStaffManager::addRod`, `InitialSpringStretch` and `GRForceRodEntry::comp` (prioritizing space rods)
- `GUIDOEngine.cpp`: `GuidoUpdateGR` toggling `checkLyricsCollisions` and calling `removeAutoSpace` when needed

Net effect
- With `checkLyricsCollisions` on: the engine performs a normal Gourlay layout, detects lyric overlaps, injects exact local spacing constraints as space rods, and re‑justifies the line.
- With it off (or when turned off later): all auto constraints are removed and the layout reverts to the pure spring–rod optimization (plus any neighborhood / Gieseking options you’ve enabled).

**Comment by basil** [comment]
The Guido strategy seems like it could work, but it is expensive. It does layout once without taking into account lyrics, then finds all the lyric collisions, then inserts Gourlay spacers for the areas where collisions occurred and does layout a second time.

More ideal would be to arrange the boxes and glue to deal with this in the initial pass, but this gets very tricky. Knuth's paper describes several cases (ragged right, ragged centered, indexes, etc) where he manipulates the box-glue construction to create specialized results, culminating in a generalized "box glue algebra" (this is peak Knuth). Key to all of this is the ability to run Knuth-Plass with negative numbers, which don't have any physical analogy but which are used to influence the Knuth Plass search in mathematical terms.

As a thought experimented, I handed `o5` the relevant portions of Knuth-Plass's paper, the Neanes source code, and asked it to come up with a boxes and glue construction. Here is the result; I have no idea if it actually works or not, but it sure sounds fancy.

```
Below the word ‘neume’ always means ❶ the graphic sign itself plus ❷ as much
horizontal room as is needed to show everything that belongs to that sign on
the present line: stems, martyria, liquescents … – but nothing that belongs to
other notes or to the syllable of another word.

Definitions that are needed once only

    s₀      = normal inter-neume distance requested by the user
    s⁺      = stretchability that may be added to s₀
    s⁻      = shrinkability that may be taken away from s₀
    ℓ       = minimal distance that two syllables must have on the same line
    ∞       = any number ≥ 10 000, i.e. “infinite” for the algorithm

The sequence that is handed to the Knuth–Plass routine is built from left to
right.  For note i we first collect the geometric data that are known after
lyrics have been attached to it.

For every note i

    Wⁿᵢ  = width of the neume itself
    Wˡᵢ  = width of the lyric that belongs to the syllable that starts here
           (0 if no syllable begins here)
    Aᵢ   = alignment of the syllable that starts here
           (left if this is the beginning of a melisma and the text is wider
           than the neume, otherwise centred)
    Lᵢ   = horizontal distance by which that syllable projects to the LEFT
           of the left edge of the printed neume (≥ 0)
    Rᵢ   = horizontal distance by which that syllable projects to the RIGHT
           of the right edge of the printed neume (≥ 0)
    Bᵢ   = width of the printable material that really has to be drawn for
           the note itself

        if Wˡᵢ ≤ Wⁿᵢ                         % syllable shorter than the sign
           Lᵢ = Rᵢ = 0                       % no projection
           Bᵢ = Wⁿᵢ
        else if Aᵢ = left                    % start of melisma
           Lᵢ = 0;  Rᵢ = Wˡᵢ – Wⁿᵢ
           Bᵢ = Wⁿᵢ
        else                                % centred syllable
           Lᵢ = Rᵢ = (Wˡᵢ – Wⁿᵢ)/2
           Bᵢ = Wⁿᵢ

    Eᵢ   = right–most x–position that is occupied by the syllable that is
           current while note i is printed.
           When a melisma continues, the number is propagated until the
           syllable really ends, possibly including the width of a printed
           hyphen.  (While scanning the music this is exactly the variable
           currentMelismaLyricsEndPx that already exists in your code.)

           On the very note where the syllable ends, Eᵢ is the right edge of
           the printed lyric; afterwards Eⱼ = 0 until the next syllable
           starts.

           We define

                       R′ᵢ = max( Rᵢ ,  Eᵢ – (left edge of note i + Wⁿᵢ) )

           I.e. R′ᵢ is the horizontal room that *must* still be kept free
           *after* the neume has been printed, no matter whether the lyric
           actually belongs to this very note or was started earlier.


Between two consecutive notes i  and i+1  we need on one and the same line

        base space         s₀
        left projection    R′ᵢ              (all projections to the right of i)
        right projection   Lᵢ₊₁            (all projections to the left of i+1)

        syllable distance  ℓ if both notes carry a syllable of their own

Put this together

        Δᵢ = R′ᵢ + Lᵢ₊₁
        ℓᵢ = ℓ if (Wˡᵢ > 0 and Wˡᵢ₊₁ > 0) else 0
        mᵢ = s₀ + Δᵢ + ℓᵢ          (minimal natural width that must be present)

How to make that width disappear completely when a line break happens right
between i and i+1?
Exactly the trick that Knuth uses for ragged-right text:

        glue( 0 ,  s⁺ ,  s⁻ )              % lives on the *left* line only
        penalty( 0 ,  0 ,  0 )             % legal break position
        glue( mᵢ , –s⁺ , –s⁻ )             % added when no break occurs

If the penalty is *not* taken we get

        0 + mᵢ         natural width  = mᵢ
        s⁺ – s⁺ = 0    stretch        = 0
        s⁻ – s⁻ = 0    shrink         = 0

exactly what is required on one line.

If the penalty *is* taken, the second glue is discarded; the previous line
receives glue (0,s⁺,s⁻) that may stretch or shrink exactly like all other
inter-neume spaces, and the new line starts flush left with note i+1.

Complete sequence for one note

Except at the very end of the paragraph every note is therefore encoded in
normal form

        %—— projections to the left ————————————————
        penalty( 0 ,  ∞ , 0 )             % never break *inside* a projection
        glue( Lᵢ , 0 , 0 )                % fixed, unbreakable

        %—— the printable sign ————————————————
        box  ( Bᵢ )
        penalty( 0 ,  ∞ , 0 )             % never break inside the sign

        %—— inter-note space that may vanish at a break ——
        glue   ( 0 , s⁺ , s⁻ )
        penalty( 0 , 0 , 0 )              % candidate breakpoint
        glue   ( mᵢ , –s⁺ , –s⁻ )

Last note of the paragraph

        penalty( 0 , ∞ , 0 )              % protect the sign itself
        glue   ( 0 , ∞ , 0 )              % the usual “finishing glue”
        penalty( 0 , –∞ , 0 )             % forced break / end of paragraph

How the four situations mentioned in the question are handled

• Syllable shorter than the neume
  ⇒  Wˡᵢ ≤ Wⁿᵢ,  therefore Lᵢ = Rᵢ = 0;  mᵢ = s₀.
  Nothing special happens.

• Syllable wider than the neume, centred
  ⇒  Lᵢ = Rᵢ > 0.
  The left half of the excess width sits in glue(Lᵢ,0,0);
  the right half is already included in mᵢ.

• Start of a melisma, syllable wider than the neume
  ⇒  Lᵢ = 0,  Rᵢ = Wˡᵢ – Wⁿᵢ.
  All additional room is put to the right side of the note, just as required.
  While the melisma continues Eⱼ is kept ≠ 0, therefore R′ⱼ remains > 0 and
  every following inter-note space contains exactly as much additional room as
  is still needed for the running syllable or for a trailing hyphen.
  If a line break occurs anywhere inside the melisma, the entire remainder of
  the syllable is moved to the next line together with the first note that
  belongs to it – exactly what a chant singer expects to see.

• Two consecutive syllables on the same line
  (the commonest reason for ℓᵢ ≠ 0)
  The first of the two inter-note glues guarantees *at least* ℓ between the
  visual ends of the two lyrics.
  If the algorithm decides to break between the syllables the penalty is
  taken, the second glue disappears, and the distance constraint is lifted
  automatically, exactly as it should be.

Relation to Hegazy/Gourlay and LilyPond

The pair

        glue( 0 , s⁺ , s⁻ )
        penalty( 0 , 0 , 0 )

is the same “pre-stretched invisible object” that Hegazy & Gourlay call a
spacing rod and that LilyPond instantiates as the `LyricSpace` or
`LyricHyphen` grob.  Its only purpose is to see that every *admissible* break
already contains the amount of stretch/shrink that the justification formula
expects, while all compulsory distances (overshoot of lyrics, minimal ℓ) are
kept in the second glue that is *removed* when the break is taken.

Putting it all together

• The sequence is purely local: only data that are known while the row of
  notes is scanned are needed.

• Every quantity that must vanish at a break is placed in the *second* glue of
  the trio  glue / penalty / glue.

• Everything that must *never* be split between two lines (part of a neume,
  part of a syllable, alignment padding on its left) is packed either into one
  box or into an ordinary glue that is protected by an infinite penalty.

Consequently the same one-pass Knuth–Plass algorithm that TeX uses for prose
now chooses musical line breaks that

  – avoid collisions between lyrics,
  – keep neighbouring syllables at least ℓ apart,
  – let ordinary inter-neume spaces stretch or shrink by ±½,
  – do not require any special treatment at the end of a melisma, and
  – cost you no more than the classical five item types:
        box,          glue,          penalty,
        ∞–penalty,    –∞–penalty.

No other primitives are required.
```

**Comment by basil** [comment]
> A good data set would be to regenerate via the command line all the scores in https://github.com/basil/byzantine-music using the latest released engine, followed by the new engine and run a diff.

@danielgarthur Yes. As described in https://github.com/neanes/neanes/issues/90#issuecomment-4129946759, you can see the results on my web site by adding "knuth-plass" to any URL. For example:

Before: https://byzantinemusic.org/tmp/Divine%20Liturgy/Leitourgika%20and%20Axion%20Estin/Axion%20Estin%201%20Slow%20Heirmologic.pdf

After: https://byzantinemusic.org/tmp/knuth-plass/Divine%20Liturgy/Leitourgika%20and%20Axion%20Estin/Axion%20Estin%201%20Slow%20Heirmologic.pdf

I've diffed the scores visually, and fixed all known bugs (dozens of them), but of course I would appreciate additional testing.

**Comment by ilizol** [comment]
> [@danielgarthur](https://github.com/danielgarthur) Yes. As described in [#90 (comment)](https://github.com/neanes/neanes/issues/90#issuecomment-4129946759), you can see the results on my web site by adding "knuth-plass" to any URL. For example:
> 
> Before: https://byzantinemusic.org/tmp/Divine%20Liturgy/Leitourgika%20and%20Axion%20Estin/Axion%20Estin%201%20Slow%20Heirmologic.pdf
> 
> After: https://byzantinemusic.org/tmp/knuth-plass/Divine%20Liturgy/Leitourgika%20and%20Axion%20Estin/Axion%20Estin%201%20Slow%20Heirmologic.pdf

@basil 
The links are broken. How we can see the "knuth-plass" results?

**Comment by basil** [comment]
I've switched my web site to use the Knuth-Plass as the default branch for all scores, since there hasn't been any motion on this PR. If someone is still interested in seeing the "old" view (current release) I suppose I could restore that somewhere.

**Comment by ilizol** [comment]
@basil 
So all the pdf files use the Knuth-Plass branch?

Saw that the 5th line starts with kentimata:
https://byzantinemusic.org/pdf/Original%20Melodies/First%20Mode/Thou%20art%20the%20joy%20Slow.pdf

**Comment by basil** [comment]
> So all the pdf files use the Knuth-Plass branch?

Yes that is correct.

> Saw that the 5th line starts with kentimata

Yes, this will still occur sometimes when no better solution can be found. Such breaks are discouraged but not banned.

**Comment by ilizol** [comment]
@basil 

Should the line start with kentimata be banned?
It is too ugly to start a line with kentimata.

https://github.com/basil/neanes/blob/knuth-plass/notes/knuth_plass.md#prohibited-breaks
https://github.com/basil/neanes/blob/knuth-plass/notes/knuth_plass.md#strongly-discouraged-breaks

**Comment by basil** [comment]
Perhaps. It does occur occasionally in classical books, but it is very, very rare. Banning the break results in a violation of the adjustment ratio and rivers of whitespace, so the alternative isn't much better, either.

The real reason the classical books don't have this problem is because they implement another set of techniques:

> To compress a line, reduced-width versions of neumes such as the ison and oligon are sometimes used.
> To compress a line, an ison followed by an apostrophos is sometimes written with the apostrophos beneath the ison rather than after it.
> To compress a line, two apostrophoi are sometimes written with the second apostrophos beneath the first rather than after it.

Implementing the above would give the solver more options to work with so that it doesn't have to break at a kentimata. But the Knuth-Plass algorithm doesn't support this today and an extension would need to be developed, similar to the one I already developed for negative widths. That is probably going to be a decent amount of work, so I am leaving that out of scope for the current effort, although I may work on it once this PR is merged.
